state-chain/pallets/cf-validator/src/lib.rs

```diff
@@ -1018,7 +1018,14 @@ impl<T: Config> Pallet<T> {
 	}
 
 	fn expire_epoch(epoch: EpochIndex) -> Weight {
-		LastExpiredEpoch::<T>::set(epoch);
+		// Recursively expire all previous epochs that have not been expired yet. This is possible
+		// due to epoch duration being configurable.
+		let previous_epoch = epoch.saturating_sub(1);
+		if epoch > 0 && LastExpiredEpoch::<T>::get() != previous_epoch {
+			Self::expire_epoch(previous_epoch);
+		}
```

Comment:
I don't see a good reason to use recursion here, much clearer to just iterate. It also breaks the semantics of the function, it's no longer 'expire epoch', it's 'expire this and any previous unexpired epoch'. So at the very least we should rename it. 

I think a better option is to take the responsibility of updating LastExpiredEpoch outside of this function to the caller.